_posts/2017-11-02-totw-101.md

```diff
@@ -0,0 +1,96 @@
+
+By Titus Winters (titus@google.com)
+
+```c++
+const string& name = obj.GetName();
+std::unique_ptr<Consumer> consumer(new Consumer(name));
+```
+
+In particular, I'd like to draw your attention to the `&`. Is it appropriate?
+What should we check? What can go wrong? I find a decent number of C++
+programmers that aren't entirely clear on references, but do generally know that
+they "avoid making copies." As with most issues in C++, it's more complicated
+than that.
+
+### Case by case: what is being returned and how is it being stored?
+
+There are two (maybe three) important questions here:
+
+1.  What type is returned (in this example, by `GetName()`)?
+1.  What type are we storing into/initializing (in this example, what is the
+    type of `name`)?
+1.  If we are returning a reference, does the object being returned by reference
+    have any lifetime limitations?
+
+We'll continue with `string` as our example type, but the same arguments
+generalize for most non-trivial value types.
+
+1.  Returning `string`, initializing `string`: This is usually RVO and is
+    guaranteed to be at worst a move for modern types (abseil.io/tips/77).
+2.  Returning `string&` or `const string&`, initializing `string`: This is a
+    copy (there must be some long-living object which we are returning a
+    reference to, so once we initialize a new string there are two names for
+    that data, hence a copy. See abseil.io/tips/77). Sometimes this is valuable,
+    like if you need your `string` to outlast the lifetime guarantees provided
+    by the function.
+3.  Returning `string`, initializing `string&`: This won't compile, as you
+    cannot bind a reference to a temporary.
+4.  Returning `const string&`, initializing `string&`: This won't compile, as
+    you've dropped the const inappropriately.
+5.  Returning `const string&`, initializing `const string&`: This is no cost
+    (you're effectively returning just a pointer). However, you have inherited
+    any existing lifetime restrictions: how long is that reference good for?
+    Most accessor methods that return a reference are returning a member, so at
+    most the reference can be valid for the life of the containing object.
+6.  Returning `string&`, initializing `string&`: This is the same as #5, but
+    with the additional caveat: the returned reference is non-const, so any
+    modifications to your reference will be reflected in the source.
+7.  Returning `string&`, initializing `const string&`: The same as #5.
+8.  Returning `string`, initializing `const string&`: You'd think, given #3,
+    that this wouldn't work. However, the language has special-case support for
+    this: if you initialize a `const T&` with a temporary `T`, that `T` (in this
+    case `string`) is not destroyed until the *reference* goes out of scope (in
+    the common case of automatic or static variables).
+
+Scenario #8 is what allows most reflexive use of references (that is "Oh, I
```

Comment:
&num; here and for #1 below.

Reply:
I don't know what you're requesting here.

Reply:
Change "#" to "&num;" entity.

\# is reserved in markdown, and may cause problems if this text is ever shifted.

Reply:
You can also escape it via \# like I did in this comment :)

Reply:
Your comments keep getting un-escaped, but hopefully done. :P 
(The difference between markdown dialects is irksome.)